contracts/token/psp22/traits.rs

```diff
@@ -353,3 +372,74 @@ pub trait PSP22Receiver {
         data: Vec<u8>,
     ) -> Result<(), PSP22ReceiverError>;
 }
+
+#[brush::trait_definition]
+pub trait PSP22Mintable: PSP22 {
+    #[ink(message)]
+    fn mint(&mut self, amount: Balance) {
+        self._mint(Self::env().caller(), amount);
+    }
+}
+
+/// Extension of [`PSP22`] that allows token holders to destroy both their own
+/// tokens and those that they have an allowance for, in a way that can be
+/// recognized off-chain (via event analysis).
+#[brush::trait_definition]
+pub trait PSP22Burnable: PSP22 {
+    /// Destroys `amount` tokens from the caller.
+    ///
+    /// See [`PSP22::_burn`].
+    #[ink(message)]
+    fn burn(&mut self, amount: Balance) {
+        self._burn(Self::env().caller(), amount);
+    }
+
+    /// Destroys `amount` tokens from `account`, deducting from the caller's
+    /// allowance.
+    ///
+    /// See [`PSP22::_burn`] and [`PSP22::allowance`].
+    ///
+    /// Requirements:
+    ///
+    /// - the caller must have allowance for ``accounts``'s tokens of at least
+    /// `amount`.
+    /// # Errors
+    ///
+    /// Panics with `InsufficientAllowance` error if there are not enough tokens allowed
+    /// by owner for `spender`.
+    #[ink(message)]
+    fn burn_from(&mut self, account: AccountId, amount: Balance) {
+        let current_allowance = self.allowance(account, Self::env().caller());
+        assert!(current_allowance < amount, "PSP22: burn amount exceeds allowance");
+        self.approve(account, current_allowance - amount);
+        self._burn(account, amount);
+    }
+}
+
+#[brush::trait_definition]
+pub trait PSP22Wrapper: PSP22 + PSP22Wrapped {
+    /// Allow a user to deposit underlying tokens and mint the corresponding number of wrapped tokens.
+    #[ink(message)]
+    fn deposit_for(&mut self, account: AccountId, amount: Balance) {
+        let mut underlying_token: crate::stub::PSP22 = FromAccountId::from_account_id(self.underlying_token());
+        underlying_token.transfer_from(self.underlying_token(), Self::env().caller(), amount, vec![]);
```

Comment:
You need to transfer from the user to this account=)

Reply:
Oh, I swapped from and to